File: app/src/main/java/com/kota/Bahamut/BahamutController.java
package com.kota.Bahamut;

import static com.kota.Bahamut.Service.CommonFunctions.changeScreenOrientation;
import static com.kota.Bahamut.Service.MyBillingClient.checkPurchase;
import static com.kota.Bahamut.Service.MyBillingClient.closeBillingClient;
import static com.kota.Bahamut.Service.MyBillingClient.initBillingClient;

import android.content.Intent;
import android.util.Log;

import com.kota.ASFramework.Dialog.ASAlertDialog;
import com.kota.ASFramework.Dialog.ASProcessingDialog;
import com.kota.ASFramework.PageController.ASNavigationController;
import com.kota.ASFramework.PageController.ASViewController;
import com.kota.ASFramework.Thread.ASRunner;
import com.kota.ASFramework.UI.ASToast;
import com.kota.Bahamut.DataModels.ArticleTempStore;
import com.kota.Bahamut.DataModels.BookmarkStore;
import com.kota.Bahamut.Pages.Model.BoardEssencePageItem;
import com.kota.Bahamut.Pages.Model.BoardPageBlock;
import com.kota.Bahamut.Pages.Model.BoardPageItem;
import com.kota.Bahamut.Pages.Model.ClassPageBlock;
import com.kota.Bahamut.Pages.Model.ClassPageItem;
import com.kota.Bahamut.Pages.Model.MailBoxPageBlock;
import com.kota.Bahamut.Pages.Model.MailBoxPageItem;
import com.kota.Bahamut.Pages.StartPage;
import com.kota.Bahamut.Pages.Theme.ThemeStore;
import com.kota.Bahamut.Service.BahaBBSBackgroundService;
import com.kota.Bahamut.Service.CloudBackup;
import com.kota.Bahamut.Service.NotificationSettings;
import com.kota.Bahamut.Service.TempSettings;
import com.kota.Bahamut.Service.UserSettings;

import android.content.pm.PackageManager;
import android.os.Build;

import com.kota.Telnet.TelnetClient;
import com.kota.Telnet.TelnetClientListener;
import com.kota.TelnetUI.TelnetPage;
import com.kota.TextEncoder.B2UEncoder;
import com.kota.TextEncoder.U2BEncoder;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;
import java.util.Vector;

public class BahamutController extends ASNavigationController implements TelnetClientListener {
    @Override // com.kota.ASFramework.PageController.ASNavigationController
    protected void onControllerWillLoad() {
        requestWindowFeature(1);
        try {
            B2UEncoder.constructInstance(getResources().openRawResource(R.raw.b2u));
            U2BEncoder.constructInstance(getResources().openRawResource(R.raw.u2b));
        } catch (Exception e) {
            Log.e(getClass().getSimpleName(), e.getMessage() != null ? e.getMessage() : "");
        }
        // 書籤
        String bookmark_file_path = getFilesDir().getPath() + "/bookmark.dat";
        BookmarkStore.upgrade(this, bookmark_file_path);

        // 暫存檔
        String article_file_path = getFilesDir().getPath() + "/article_temp.dat";
        ArticleTempStore.upgrade(this, article_file_path);

        // 系統架構
        TelnetClient.construct(BahamutStateHandler.getInstance());
        TelnetClient.getClient().setListener(this);

        // 設定 TelnetConnector 的設備控制器
        TelnetClient.getConnector().setDeviceController(getDeviceController());

        PageContainer.constructInstance();

        // UserSettings
        setAnimationEnable(UserSettings.getPropertiesAnimationEnable());
        // 啟用wifi鎖定
        if (UserSettings.getPropertiesKeepWifi()) {
            getDeviceController().lockWifi();
        }

        // 共用函數
        TempSettings.myContext = this;
        TempSettings.myActivity = ASNavigationController.getCurrentController();
        changeScreenOrientation();

        // 以下需等待 共用函數 設定完畢
        // VIP
        TempSettings.applicationContext = getApplicationContext();
        initBillingClient();

        // 外觀
        ThemeStore.INSTANCE.upgrade(this);
    }

    @Override // com.kota.ASFramework.PageController.ASNavigationController
    protected void onControllerDidLoad() {
        StartPage start_page = PageContainer.getInstance().getStartPage();
        pushViewController(start_page, false);
    }

    @Override
    protected void onResume() {
        checkPurchase();

        super.onResume();
    }

    @Override
    protected void onDestroy() {
        // 關閉VIP
        closeBillingClient();

        // 備份雲端
        if (NotificationSettings.getCloudSave()) {
            CloudBackup cloudBackup = new CloudBackup();
            cloudBackup.backup();
        }

        // 強制關閉連線
        TelnetClient.getClient().close();

        // 清理 TelnetConnector 的設備控制器引用
        if (TelnetClient.getConnector() != null) {
            TelnetClient.getConnector().setDeviceController(null);
        }

        super.onDestroy();
    }

    @Override // com.kota.ASFramework.PageController.ASNavigationController,
              // android.app.Activity
    protected void onPause() {
        super.onPause();
    }

    @Override // com.kota.ASFramework.PageController.ASNavigationController
    protected String getControllerName() {
        return String.valueOf(R.string.app_name);
    }

    @Override // com.kota.ASFramework.PageController.ASNavigationController
    public boolean onBackLongPressed() {
        boolean result = true;
        if (TelnetClient.getConnector().isConnecting()) {
            ASAlertDialog dialog = new ASAlertDialog();
            // from class: com.kota.Bahamut.BahamutController.1
// com.kota.ASFramework.Dialog.ASAlertDialogListener
            dialog
                    .setMessage("是否確定要強制斷線?")
                    .addButton("取消")
                    .addButton("斷線")
                    .setListener((aDialog, index) -> {
                        if (index == 1) {
                            TelnetClient.getClient().close();
                        }
                    }).show();
        } else {
            result = false;
        }
        ASProcessingDialog.dismissProcessingDialog();
        return result;
    }

    private void showConnectionStartMessage() {
        SimpleDateFormat date_format = new SimpleDateFormat("yyyy-MM-dd kk:hh:ss", Locale.TRADITIONAL_CHINESE);
        date_format.setTimeZone(TimeZone.getTimeZone("GMT+8"));
        String time_string = date_format.format(new Date());
        System.out.println("BahaBBS connection start:" + time_string);
    }

    @Override // com.kota.Telnet.TelnetClientListener
    public void onTelnetClientConnectionStart(TelnetClient aClient) {
        new ASRunner() { // from class: com.kota.Bahamut.BahamutController.2
            @Override // com.kota.ASFramework.Thread.ASRunner
            public void run() {
                BahamutController.this.showConnectionStartMessage();
            }
        }.runInMainThread();
    }

    @Override // com.kota.Telnet.TelnetClientListener
    public void onTelnetClientConnectionSuccess(TelnetClient aClient) {
        // 先檢查通知權限
        boolean isPermissionGranted = checkNotificationPermissionBeforeStartService();

        if (isPermissionGranted) {
            try {
                Intent intent = new Intent(this, BahaBBSBackgroundService.class);
                startForegroundService(intent);
            } catch (SecurityException e) {
                // Handle permission issues for foreground service
                Log.e(getClass().getSimpleName(), "Failed to start foreground service: SecurityException", e);
                fallbackToRegularService();
            } catch (IllegalStateException e) {
                // Handle cases where the app is in background and can't start foreground
                // service
                Log.e(getClass().getSimpleName(), "Failed to start foreground service: IllegalStateException", e);
                fallbackToRegularService();
            } catch (Exception e) {
                // Handle ForegroundServiceStartNotAllowedException and other exceptions
                Log.e(getClass().getSimpleName(), "Failed to start foreground service: " + e.getClass().getSimpleName(),
                        e);
                fallbackToRegularService();
            }
        }
    }

    /**
     * 當前景服務啟動失敗時，嘗試啟動一般服務作為備用方案
     */
    private void fallbackToRegularService() {
        try {
            Intent intent = new Intent(this, BahaBBSBackgroundService.class);
            startService(intent);
            Log.i(getClass().getSimpleName(), "Fallback to regular service successful");
        } catch (Exception fallbackException) {
            Log.e(getClass().getSimpleName(), "Failed to start service as fallback", fallbackException);
        }
    }

    /**
     * 安全地停止背景服務，避免 ForegroundServiceDidNotStopInTimeException
     */
    private void stopBackgroundServiceSafely() {
        try {
            Intent intent = new Intent(this, BahaBBSBackgroundService.class);
            boolean result = stopService(intent);
            Log.i(getClass().getSimpleName(), "Stop service result: " + result);
        } catch (Exception e) {
            Log.e(getClass().getSimpleName(), "Error stopping background service: " + e.getClass().getSimpleName(), e);
        }
    }

    @Override // com.kota.Telnet.TelnetClientListener
    public void onTelnetClientConnectionFail(TelnetClient aClient) {
        ASProcessingDialog.dismissProcessingDialog();
        ASToast.showShortToast("連線失敗，請檢查網路連線或稍後再試");
    }

    @Override // com.kota.Telnet.TelnetClientListener
    public void onTelnetClientConnectionClosed(TelnetClient aClient) {
        // 優雅地停止背景服務
        stopBackgroundServiceSafely();
        new ASRunner() { // from class: com.kota.Bahamut.BahamutController.3
            @Override // com.kota.ASFramework.Thread.ASRunner
            public void run() {
                SimpleDateFormat date_format = new SimpleDateFormat("yyyy-MM-dd kk:hh:ss", Locale.TRADITIONAL_CHINESE);
                date_format.setTimeZone(TimeZone.getTimeZone("GMT+8"));
                String time_string = date_format.format(new Date());
                System.out.println("BahaBBS connection close:" + time_string);

                BahamutController.this.handleNormalConnectionClosed();

                ASToast.showShortToast("連線已中斷");

                ASProcessingDialog.dismissProcessingDialog();

                if (NotificationSettings.getCloudSave()) {
                    CloudBackup cloudBackup = new CloudBackup();
                    cloudBackup.backup();
                }

                if (TempSettings.getMessageSmall() != null) {
                    getCurrentController().removeForeverView(TempSettings.getMessageSmall());
                    TempSettings.setMessageSmall(null);
                }
            }
        }.runInMainThread();
    }

    private void handleNormalConnectionClosed() {
        Vector<ASViewController> pages = ASNavigationController.getCurrentController().getViewControllers();
        Vector<ASViewController> new_controllers = new Vector<>();
        for (ASViewController controller : pages) {
            TelnetPage telnet_page = (TelnetPage) controller;
            if (telnet_page.getPageType() == 0 || telnet_page.isKeepOnOffline()) {
                new_controllers.add(telnet_page);
            }
        }
        StartPage start_page = PageContainer.getInstance().getStartPage();
        if (!new_controllers.contains(start_page)) {
            new_controllers.insertElementAt(start_page, 0);
        }
        setViewControllers(new_controllers);
    }

    @Override // com.kota.ASFramework.PageController.ASNavigationController,
              // android.app.Activity, android.content.ComponentCallbacks
    public void onLowMemory() {
        super.onLowMemory();
        BoardPageBlock.release();
        BoardPageItem.release();
        BoardEssencePageItem.Companion.release();
        ClassPageBlock.release();
        ClassPageItem.release();
        MailBoxPageBlock.release();
        MailBoxPageItem.release();
        System.gc();
    }

    @Override // com.kota.ASFramework.PageController.ASNavigationController
    public boolean isAnimationEnable() {
        return UserSettings.getPropertiesAnimationEnable();
    }

    /**
     * 在啟動服務前檢查通知權限
     */
    private boolean checkNotificationPermissionBeforeStartService() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            if (checkSelfPermission(android.Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) {
                Log.w(getClass().getSimpleName(), "POST_NOTIFICATIONS permission not granted");
                return false;
            }
        }
        return true;
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
    }
}

File: app/src/main/java/com/kota/Bahamut/BahamutController.kt
package com.kota.Bahamut

import android.content.Intent
import android.util.Log
import com.kota.Bahamut.dataModels.ArticleTempStore
import com.kota.Bahamut.dataModels.BookmarkStore
import com.kota.Bahamut.pages.StartPage
import com.kota.Bahamut.pages.messages.MessageSmall
import com.kota.Bahamut.pages.model.BoardEssencePageItem
import com.kota.Bahamut.pages.model.BoardPageBlock
import com.kota.Bahamut.pages.model.BoardPageItem
import com.kota.Bahamut.pages.model.ClassPageBlock
import com.kota.Bahamut.pages.model.ClassPageItem
import com.kota.Bahamut.pages.model.MailBoxPageBlock
import com.kota.Bahamut.pages.model.MailBoxPageItem
import com.kota.Bahamut.pages.theme.ThemeStore.upgrade
import com.kota.Bahamut.service.BahaBBSBackgroundService
import com.kota.Bahamut.service.CloudBackup
import com.kota.Bahamut.service.CommonFunctions.changeScreenOrientation
import com.kota.Bahamut.service.MyBillingClient.checkPurchase
import com.kota.Bahamut.service.MyBillingClient.closeBillingClient
import com.kota.Bahamut.service.MyBillingClient.initBillingClient
import com.kota.Bahamut.service.NotificationSettings.getCloudSave
import com.kota.Bahamut.service.TempSettings
import com.kota.Bahamut.service.TempSettings.getMessageSmall
import com.kota.Bahamut.service.TempSettings.setMessageSmall
import com.kota.Bahamut.service.UserSettings.Companion.propertiesAnimationEnable
import com.kota.Bahamut.service.UserSettings.Companion.propertiesKeepWifi
import com.kota.asFramework.dialog.ASAlertDialog
import com.kota.asFramework.dialog.ASAlertDialogListener
import com.kota.asFramework.dialog.ASProcessingDialog.Companion.dismissProcessingDialog
import com.kota.asFramework.pageController.ASNavigationController
import com.kota.asFramework.pageController.ASViewController
import com.kota.asFramework.thread.ASRunner
import com.kota.asFramework.ui.ASToast.showShortToast
import com.kota.telnet.TelnetClient
import com.kota.telnet.TelnetClient.Companion.construct
import com.kota.telnet.TelnetClientListener
import com.kota.telnetUI.TelnetPage
import com.kota.textEncoder.B2UEncoder
import com.kota.textEncoder.U2BEncoder
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import java.util.Vector

class BahamutController : ASNavigationController(), TelnetClientListener {
    // com.kota.asFramework.pageController.ASNavigationController
    override fun onControllerWillLoad() {
        requestWindowFeature(1)
        try {
            B2UEncoder.constructInstance(resources.openRawResource(R.raw.b2u))
            U2BEncoder.constructInstance(resources.openRawResource(R.raw.u2b))
        } catch (e: Exception) {
            Log.e(javaClass.simpleName, (if (e.message != null) e.message else "")!!)
        }
        // 書籤
        val bookmarkFilePath = filesDir.path + "/bookmark.dat"
        BookmarkStore.upgrade(this, bookmarkFilePath)

        // 暫存檔
        val articleFilePath = filesDir.path + "/article_temp.dat"
        ArticleTempStore.upgrade(this, articleFilePath)

        // 系統架構
        construct(BahamutStateHandler.Companion.getInstance())
        TelnetClient.myInstance!!.setListener(this)


        // 設定 TelnetConnector 的設備控制器
        TelnetClient.myInstance!!.telnetConnector?.setDeviceController(deviceController)

        PageContainer.Companion.constructInstance()

        // UserSettings
        isAnimationEnable = propertiesAnimationEnable
        // 啟用wifi鎖定
        if (propertiesKeepWifi) {
            deviceController?.lockWifi()
        }

        // 共用函數
        TempSettings.myContext = this
        TempSettings.myActivity = currentController
        changeScreenOrientation()

        // 以下需等待 共用函數 設定完畢
        // VIP
        TempSettings.applicationContext = applicationContext
        initBillingClient()

        // 外觀
        upgrade(this)
    }

    // com.kota.asFramework.pageController.ASNavigationController
    override fun onControllerDidLoad() {
        val startPage: StartPage? = PageContainer.instance!!.startPage
        pushViewController(startPage, false)
    }

    override fun onResume() {
        checkPurchase()
        super.onResume()
    }

    override fun onDestroy() {
        // 關閉VIP
        closeBillingClient()

        // 備份雲端
        if (getCloudSave()) {
            val cloudBackup = CloudBackup()
            cloudBackup.backup()
        }

        // 強制關閉連線
        TelnetClient.myInstance!!.close()


        // 清理 TelnetConnector 的設備控制器引用
        if (TelnetClient.myInstance!!.telnetConnector != null) {
            TelnetClient.myInstance!!.telnetConnector?.setDeviceController(null)
        }

        super.onDestroy()
    }

    override fun onPause() {
        super.onPause()
    }

    override val controllerName: String?
        // com.kota.asFramework.pageController.ASNavigationController
        get() = R.string.app_name.toString()

    // com.kota.asFramework.pageController.ASNavigationController
    override fun onBackLongPressed(): Boolean {
        var result = true
        if (TelnetClient.myInstance!!.telnetConnector?.isConnecting == true) {
            val dialog = ASAlertDialog()
            dialog
                .setMessage("是否確定要強制斷線?")
                .addButton("取消")
                .addButton("斷線")
                .setListener(object : ASAlertDialogListener {
                    // from class: com.kota.Bahamut.BahamutController.1
                    // com.kota.asFramework.dialog.ASAlertDialogListener
                    override fun onAlertDialogDismissWithButtonIndex(
                        paramASAlertDialog: ASAlertDialog,
                        paramInt: Int
                    ) {
                        if (paramInt == 1) {
                            TelnetClient.myInstance!!.close()
                        }
                    }
                }).show()
        } else {
            result = false
        }
        dismissProcessingDialog()
        return result
    }

    private fun showConnectionStartMessage() {
        val dateFormat = SimpleDateFormat("yyyy-MM-dd kk:hh:ss", Locale.TRADITIONAL_CHINESE)
        dateFormat.timeZone = TimeZone.getTimeZone("GMT+8")
        val timeString = dateFormat.format(Date())
        println("BahaBBS connection start:$timeString")
    }

    // com.kota.telnet.TelnetClientListener
    override fun onTelnetClientConnectionStart(telnetClient: TelnetClient) {
        object : ASRunner() {
            // from class: com.kota.Bahamut.BahamutController.2
            // com.kota.asFramework.thread.ASRunner
            override fun run() {
                this@BahamutController.showConnectionStartMessage()
            }
        }.runInMainThread()
    }

    // com.kota.telnet.TelnetClientListener
    override fun onTelnetClientConnectionSuccess(telnetClient: TelnetClient) {
        val intent = Intent(this, BahaBBSBackgroundService::class.java)
        startForegroundService(intent)
    }

    // com.kota.telnet.TelnetClientListener
    override fun onTelnetClientConnectionFail(telnetClient: TelnetClient) {
        dismissProcessingDialog()
        showShortToast("連線失敗，請檢查網路連線或稍後再試")
    }

    // com.kota.telnet.TelnetClientListener
    override fun onTelnetClientConnectionClosed(telnetClient: TelnetClient) {
        val intent = Intent(this, BahaBBSBackgroundService::class.java)
        stopService(intent)
        object : ASRunner() {
            override fun run() {
                val dateFormat = SimpleDateFormat("yyyy-MM-dd kk:hh:ss", Locale.TRADITIONAL_CHINESE)
                dateFormat.timeZone = TimeZone.getTimeZone("GMT+8")
                val timeString = dateFormat.format(Date())
                println("BahaBBS connection close:$timeString")

                this@BahamutController.handleNormalConnectionClosed()

                showShortToast("連線已中斷")

                dismissProcessingDialog()

                if (getCloudSave()) {
                    val cloudBackup = CloudBackup()
                    cloudBackup.backup()
                }

                if (getMessageSmall() != null) {
                    val messageSmall: MessageSmall? = getMessageSmall()
                    currentController?.removeForeverView(messageSmall)
                    setMessageSmall(null)
                }
            }
        }.runInMainThread()
    }

    private fun handleNormalConnectionClosed() {
        val pages: Vector<ASViewController> = currentController!!.viewControllers
        val newControllers = Vector<ASViewController>()
        for (controller in pages) {
            val telnetPage = controller as TelnetPage
            if (telnetPage.pageType == 0 || telnetPage.isKeepOnOffline) {
                newControllers.add(telnetPage)
            }
        }
        val startPage: StartPage? = PageContainer.instance!!.startPage
        if (!newControllers.contains(startPage)) {
            newControllers.insertElementAt(startPage, 0)
        }
        setViewControllers(newControllers)
    }

    // com.kota.asFramework.pageController.ASNavigationController, android.app.Activity, android.content.ComponentCallbacks
    override fun onLowMemory() {
        super.onLowMemory()
        BoardPageBlock.release()
        BoardPageItem.release()
        BoardEssencePageItem.Companion.release()
        ClassPageBlock.release()
        ClassPageItem.release()
        MailBoxPageBlock.release()
        MailBoxPageItem.release()
        System.gc()
    }

    override var isAnimationEnable: Boolean = false
        // com.kota.asFramework.pageController.ASNavigationController
        get() = propertiesAnimationEnable

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
    }
}
